File: Cap3/spline.py
import numpy as np
import matplotlib.pyplot as plt

def Spline(x, y, d):
    n = len(x)
    A = np.zeros(((d + 1) * (n - 1), (d + 1) * (n - 1)))
    b = np.zeros((d + 1) * (n - 1))
    cua = x ** 2
    cub = x ** 3
    
    # Spline Lineal
    if d == 1:
        c, h = 0, 0
        for i in range(n - 1):
            A[h, c] = x[i]
            A[h, c + 1] = 1
            b[h] = y[i]
            c += 2
            h += 1
        
        c = 0
        for i in range(1, n):
            A[h, c] = x[i]
            A[h, c + 1] = 1
            b[h] = y[i]
            c += 2
            h += 1

    # Spline Cuadrático
    elif d == 2:
        c, h = 0, 0
        for i in range(n - 1):
            A[h, c] = cua[i]
            A[h, c + 1] = x[i]
            A[h, c + 2] = 1
            b[h] = y[i]
            c += 3
            h += 1
        
        c = 0
        for i in range(1, n):
            A[h, c] = cua[i]
            A[h, c + 1] = x[i]
            A[h, c + 2] = 1
            b[h] = y[i]
            c += 3
            h += 1
        
        c = 0
        for i in range(1, n - 1):
            A[h, c] = 2 * x[i]
            A[h, c + 1] = 1
            A[h, c + 3] = -2 * x[i]
            A[h, c + 4] = -1
            b[h] = 0
            c += 3
            h += 1
        
        A[h, 0] = 2
        b[h] = 0

    # Spline Cúbico
    elif d == 3:
        c, h = 0, 0
        for i in range(n - 1):
            A[h, c] = cub[i]
            A[h, c + 1] = cua[i]
            A[h, c + 2] = x[i]
            A[h, c + 3] = 1
            b[h] = y[i]
            c += 4
            h += 1

        c = 0
        for i in range(1, n):
            A[h, c] = cub[i]
            A[h, c + 1] = cua[i]
            A[h, c + 2] = x[i]
            A[h, c + 3] = 1
            b[h] = y[i]
            c += 4
            h += 1

        c = 0
        for i in range(1, n - 1):
            A[h, c] = 3 * cua[i]
            A[h, c + 1] = 2 * x[i]
            A[h, c + 2] = 1
            A[h, c + 4] = -3 * cua[i]
            A[h, c + 5] = -2 * x[i]
            A[h, c + 6] = -1
            b[h] = 0
            c += 4
            h += 1
        
        c = 0
        for i in range(1, n - 1):
            A[h, c] = 6 * x[i]
            A[h, c + 1] = 2
            A[h, c + 4] = -6 * x[i]
            A[h, c + 5] = -2
            b[h] = 0
            c += 4
            h += 1
        
        A[h, 0] = 6 * x[0]
        A[h, 1] = 2
        b[h] = 0
        h += 1
        A[h, c] = 6 * x[-1]
        A[h, c + 1] = 2
        b[h] = 0

    # Resolución del sistema
    val = np.linalg.solve(A, b)
    Tabla = val.reshape(n - 1, d + 1)

    # Gráfica de los polinomios spline
    plt.figure()
    for i in range(n - 1):
        coef = Tabla[i, :]
        x_vals = np.linspace(x[i], x[i + 1], 100)
        
        if d == 1:
            y_vals = coef[0] * x_vals + coef[1]
        elif d == 2:
            y_vals = coef[0] * x_vals**2 + coef[1] * x_vals + coef[2]
        elif d == 3:
            y_vals = coef[0] * x_vals**3 + coef[1] * x_vals**2 + coef[2] * x_vals + coef[3]
        
        plt.plot(x_vals, y_vals, 'b', linewidth=1.5)
    
    plt.plot(x, y, 'ro', markerfacecolor='r')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.title(f'Spline de Interpolación (Grado {d})')
    plt.grid()
    plt.show()

    return Tabla

File: Cap3/Spline.m
%Spline: Calcula los coeficientes de los polinomios de interpolación de
% grado d (1, 2, 3) para el conjunto de n datos (x,y), 
% mediante el método spline.
function [Tabla] = Spline(x, y, d)
    n = length(x);
    A = zeros((d + 1) * (n - 1));
    b = zeros((d + 1) * (n - 1), 1);
    cua = x .^ 2;
    cub = x .^ 3;
    
    %% lineal
    if d == 1
        c = 1;
        h = 1;
        for i = 1:n-1
            A(i, c) = x(i);
            A(i, c + 1) = 1;
            b(i) = y(i);
            c = c + 2;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n
            A(h, c) = x(i);
            A(h, c + 1) = 1;
            b(h) = y(i);
            c = c + 2;
            h = h + 1;
        end
        
    %% Cuadratic
    elseif d == 2
        c = 1;
        h = 1;
        for i = 1:n-1
            A(i, c) = cua(i);
            A(i, c + 1) = x(i);
            A(i, c + 2) = 1;
            b(i) = y(i);
            c = c + 3;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n
            A(h, c) = cua(i);
            A(h, c + 1) = x(i);
            A(h, c + 2) = 1;
            b(h) = y(i);
            c = c + 3;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n-1
            A(h, c) = 2 * x(i);
            A(h, c + 1) = 1;
            A(h, c + 3) = -2 * x(i);
            A(h, c + 4) = -1;
            b(h) = 0;
            c = c + 4;
            h = h + 1;
        end
        
        A(h, 1) = 2;
        b(h) = 0;
        
    %% Cubic
    elseif d == 3
        c = 1;
        h = 1;
        for i = 1:n-1
            A(i, c) = cub(i);
            A(i, c + 1) = cua(i);
            A(i, c + 2) = x(i);
            A(i, c + 3) = 1;
            b(i) = y(i);
            c = c + 4;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n
            A(h, c) = cub(i);
            A(h, c + 1) = cua(i);
            A(h, c + 2) = x(i);
            A(h, c + 3) = 1;
            b(h) = y(i);
            c = c + 4;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n-1
            A(h, c) = 3 * cua(i);
            A(h, c + 1) = 2 * x(i);
            A(h, c + 2) = 1;
            A(h, c + 4) = -3 * cua(i);
            A(h, c + 5) = -2 * x(i);
            A(h, c + 6) = -1;
            b(h) = 0;
            c = c + 4;
            h = h + 1;
        end
        
        c = 1;
        for i = 2:n-1
            A(h, c) = 6 * x(i);
            A(h, c + 1) = 2;
            A(h, c + 4) = -6 * x(i);
            A(h, c + 5) = -2;
            b(h) = 0;
            c = c + 4;
            h = h + 1;
        end
        
        A(h, 1) = 6 * x(1);
        A(h, 2) = 2;
        b(h) = 0;
        h = h + 1;
        A(h, c) = 6 * x(end);
        A(h, c + 1) = 2;
        b(h) = 0;
    end

    val = inv(A) * b;
    Tabla = reshape(val, d + 1, n - 1)';
    
    %% Graficar los polinomios spline
    hold on;
    for i = 1:n-1
        % Obtener los coeficientes de cada polinomio
        coef = Tabla(i, :);
        
        % Crear puntos en el intervalo [x(i), x(i+1)]
        x_vals = linspace(x(i), x(i + 1), 100);
        
        % Evaluar el polinomio en x_vals
        if d == 1
            y_vals = coef(1) * x_vals + coef(2);
        elseif d == 2
            y_vals = coef(1) * x_vals.^2 + coef(2) * x_vals + coef(3);
        elseif d == 3
            y_vals = coef(1) * x_vals.^3 + coef(2) * x_vals.^2 + coef(3) * x_vals + coef(4);
        end
        
        % Graficar el tramo del spline
        plot(x_vals, y_vals, 'b', 'LineWidth', 1.5);
    end
    
    % Graficar los puntos originales
    plot(x, y, 'ro', 'MarkerFaceColor', 'r');

    xlabel('x');
    ylabel('y');
    title('Spline de Interpolación');
    grid on;
    hold off;
end
